src/freenet/clients/fcp/FCPPluginClient.java

```diff
@@ -0,0 +1,1190 @@
+     *            cause deadlocks on its own, but you might produce deadlocks with your own thread
+     *            synchronization measures.<br><br>
+     * 
+     * @param message
+     *            You <b>must not</b> send the same message twice: This can break
+     *            {@link #sendSynchronous(SendDirection, FCPPluginMessage, long)}.<br>
+     *            To ensure this, always construct a fresh FCPPluginMessage object when re-sending
+     *            a message. If you use the constructor which allows specifying your own identifier,
+     *            always generate a fresh, random identifier.<br>
+     *            TODO: Code quality: Add a flag to FCPPluginMessage which marks the message as
+     *            sent and use it to log an error if someone tries to send the same message twice.
+     *            <br><br>
+     * 
+     * @throws IOException
+     *             If the connection has been closed meanwhile.<br/>
+     *             This FCPPluginClient <b>should be</b> considered as dead once this happens, you
+     *             should then discard it and obtain a fresh one.
+     * 
+     *             <p><b>ATTENTION:</b> If this is not thrown, that does NOT mean that the
+     *             connection is alive. Messages are sent asynchronously, so it can happen that a
+     *             closed connection is not detected before this function returns.<br/>
+     *             The only way of knowing that a send succeeded is by receiving a reply message
+     *             in your {@link FredPluginFCPMessageHandler}.<br>
+     *             If you need to know whether the send succeeded on the same thread which shall
+     *             call the send function, you can also use {@link #sendSynchronous(SendDirection,
+     *             FCPPluginMessage, long)} which will return the reply right away.</p>
+     * @see #sendSynchronous(SendDirection, FCPPluginMessage, long)
+     *          You may instead use the blocking sendSynchronous() if your thread needs to know
+     *          whether messages arrived, to ensure a certain order of arrival, or to know
+     *          the reply to a message.
+     */
+    public void send(final SendDirection direction, FCPPluginMessage message) throws IOException {
+        // We first have to compute the message.permissions field ourselves - we shall ignore what
+        // caller said for security.
+        ClientPermissions currentPermissions = (direction == SendDirection.ToClient) ?
+             null // Server-to-client messages do not have permissions.
+             : computePermissions();
+
+        // We set the permissions by creating a fresh FCPPluginMessage object so the caller cannot
+        // overwrite what we compute.
+        message = FCPPluginMessage.constructRawMessage(currentPermissions, message.identifier,
+            message.params, message.data, message.success, message.errorCode, message.errorMessage);
```

Comment:
I would like to keep the amount of internal functions in this class as low as possible:
- It aims at being read by all plugin authors before writing a plugin, as an overview for what fields a FCP plugin network message provides
- It is already rather large.
- Using package private visibility is not an option since it will be in the freenet.pluginmanager package.

ACK to keep as is?